mule-extensions-api/src/main/java/org/mule/runtime/extension/api/client/ExtensionsClient.java

```diff
@@ -0,0 +1,86 @@
+ * need of much manual coding and allowing to scale quickly all of this without a the need of compile dependency to
+ * the executed extension.
+ * <p>
+ * This API is simple and easy to use, the user just needs to define the extension and the corresponding operation
+ * that want's to invoke and pass a set of named parameters that will be attached when the operation gets executed. This implies
+ * that the user should know the parameter names and types to create them using an {@link OperationParameters} instance. The idea
+ * is that different extensions can provide different {@link OperationParameters} implementations for key operations.
+ * <p>
+ * This client lets the user reference complex configurations defined in the application and also has the capability to resolve
+ * expressions.
+ * <p>
+ * Note that this client will be reachable through the mule registry and you will be able to inject it in any class with lifecycle.
+ * <p>
+ * An usage example for an operation with this signature {@code public String getName(@UseConfig config, int account)} could be:
+ * <pre>
+ * {@code
+ * public class UsingExtensionsClient {
+ *  @Inject ExtensionsClient client;
+ *  ...
+ *  public void executeWithClient() {
+ *    OperationParameters parameters = DefaultOperationParameters.builder().configName("conf").addParameter("account", 12).build();
+ *    Result<String, Attributes> result = client.execute(CustomExtension.class, "getName", parameters);
+ *    if (result.getOutput().equals("DeveloperAccount"))
+ *    ...
+ *  }
+ * }
+ * }
+ * </pre>
+ *
+ * @since 1.0
+ */
+public interface ExtensionsClient {
+
+  /**
+   * Executes an operation asynchronously by returning a {@link Publisher} instance through which the user can obtain the
+   * result without blocking.
+   * <p>
+   * By subscribing to the returned {@link Publisher} the user can cover all the cases of processing, error and completion of the
+   * data stream.
+   * <p>
+   * When the executed operation is non-blocking, this is the recommended method to use
+   * <p>
+   * If the executed operation is not asynchronous in nature, the client might choose to actually execute in a synchronous manner.
+   *
+   * @param extension  the name of the extension that contains the operation to be executed.
+   * @param operation  the name of the operation to be executed.
+   * @param parameters an {@link OperationParameters} instance with all the parameters required to execute the operation.
+   * @return a {@link Publisher} of {@link Result}s with the payload content and the corresponding attributes to work with
+   * the operation execution asynchronously.
+   * @throws MuleException in any error case except for operation execution errors, those are handled with the {@link Publisher}.
+   */
+  <T, A extends Attributes> Publisher<Result<T, A>> executeAsync(String extension, String operation,
```

Comment:
CompletableFuture is a no go. That API is not good for this case since it's design for both dispatching the task and listening for its completion. This requires something which only provides the listening semantics. A Guava ListeningFuture would be a better fit but we shouldn't depend on Guava either.

I personally don't think that exposing a Publisher is such a big deal since it's a standard JDK9 thing now. On the contrary, I could make the case that this makes it easier for users to use this on a reactive stream. That being said, I'm open to the idea of adding an `AsyncResult` kind of class in Ext-Api or Mule-Api.

@juandesi what's your take on this?